### PR TITLE
IGNITE-19422 Fixed "get" method in index storages.

### PR DESCRIPTION
**Pull request by ibessonov**
https://issues.apache.org/jira/browse/IGNITE-19422

### CODE REVIEW

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/AbstractPageMemoryIndexStorage.java [hunk 1]
```diff
@@ -192,6 +196,123 @@ public void finishCleanup() {
...
+    /**
+     * Cursor that always returns up-to-date next element.
+     *
+     * @param <R> Type of the returned value.
+     */
+    protected abstract class ScanCursor<R> implements PeekCursor<R> {
+        private final BplusTree<K, V> indexTree;
+
+        @Nullable
+        private Boolean hasNext;
```
Comment (tkalkirill) [inline modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/AbstractPageMemoryIndexStorage.java:211]:
```suggestion
        private @Nullable Boolean hasNext;
```

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/AbstractPageMemoryIndexStorage.java [hunk 2]
```diff
@@ -192,6 +196,123 @@ public void finishCleanup() {
...
+
+        @Nullable
+        private final K lower;
```
Comment (tkalkirill) [inline modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/AbstractPageMemoryIndexStorage.java:214]:
```suggestion
        private final @Nullable K lower;
```

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/AbstractPageMemoryIndexStorage.java [hunk 3]
```diff
@@ -192,6 +196,123 @@ public void finishCleanup() {
...
+
+        @Nullable
+        private V treeRow;
```
Comment (tkalkirill) [inline modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/AbstractPageMemoryIndexStorage.java:217]: Missing javadoc

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/AbstractPageMemoryIndexStorage.java [hunk 4]
```diff
@@ -192,6 +196,123 @@ public void finishCleanup() {
...
+            }
+
+            if (peekedRow != null && halt(peekedRow)) {
+                peekedRow = null;
+            }
+
+            return peekedRow;
+        }
+
+        private boolean advanceIfNeeded() throws IgniteInternalCheckedException {
```
Comment (tkalkirill) [inline modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/AbstractPageMemoryIndexStorage.java:301]:
```suggestion
        private boolean advanceIfNeededBusy() throws IgniteInternalCheckedException {
```

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/index/AbstractRocksDbIndexStorage.java [hunk 5]
```diff
@@ -201,4 +214,125 @@ String createStorageInfo() {
...
+    /**
+     * Cursor that always returns up-to-date next element.
+     */
+    protected final class UpToDatePeekCursor<T> implements PeekCursor<T> {
+        private final Slice upperBoundSlice;
+        private final byte[] lowerBound;
+
+        private final ReadOptions options;
+        private final RocksIterator it;
+        private final Function<ByteBuffer, T> mapper;
```
Comment (tkalkirill) [inline modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/index/AbstractRocksDbIndexStorage.java:227]: https://memesmix.net/media/created/250/k5md75.jpg

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/index/AbstractRocksDbIndexStorage.java [hunk 6]
```diff
@@ -201,4 +214,125 @@ String createStorageInfo() {
...
+
+        @Nullable
+        private Boolean hasNext;
+
+        private byte @Nullable [] key;
+
+        private byte @Nullable [] peekedKey = BYTE_EMPTY_ARRAY;
```
Comment (tkalkirill) [inline modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/index/AbstractRocksDbIndexStorage.java:234]: Missing javadoc

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/index/AbstractRocksDbIndexStorage.java [hunk 7]
```diff
@@ -201,4 +214,125 @@ String createStorageInfo() {
...
+
+                if (res == null) {
+                    return null;
+                } else {
+                    return mapper.apply(ByteBuffer.wrap(res).order(KEY_BYTE_ORDER));
+                }
+            });
+        }
+
+        private byte @Nullable [] peek0() {
```
Comment (tkalkirill) [inline modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/index/AbstractRocksDbIndexStorage.java:287]:
```suggestion
        private byte @Nullable [] peekBusy() {
```

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/index/AbstractRocksDbIndexStorage.java [hunk 8]
```diff
@@ -201,4 +214,125 @@ String createStorageInfo() {
...
+
+                peekedKey = null;
+            } else {
+                peekedKey = it.key();
+            }
+
+            return peekedKey;
+        }
+
+        private boolean advanceIfNeeded() throws StorageException {
```
Comment (tkalkirill) [inline modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/index/AbstractRocksDbIndexStorage.java:305]:
```suggestion
        private boolean advanceIfNeededBusy() throws StorageException {
```

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/index/AbstractRocksDbIndexStorage.java [hunk 9]
```diff
@@ -201,4 +214,125 @@ String createStorageInfo() {
...
+
+            //noinspection ArrayEquality
+            key = (peekedKey == BYTE_EMPTY_ARRAY) ? peek0() : peekedKey;
+            peekedKey = BYTE_EMPTY_ARRAY;
+
+            hasNext = key != null;
+            return hasNext;
+        }
+
+        private void refreshAndPrepareRocksIterator() {
```
Comment (tkalkirill) [inline modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/index/AbstractRocksDbIndexStorage.java:316]:
```suggestion
        private void refreshAndPrepareRocksIteratorBusy() {
```

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/AbstractPageMemoryIndexStorage.java [hunk 10]
```diff
@@ -192,6 +196,123 @@ public void finishCleanup() {
         state.compareAndSet(StorageState.CLEANUP, StorageState.RUNNABLE);
     }
 
+    /** Constant that represents the absence of value in {@link ScanCursor}. */
+    private static final IndexRowKey NO_INDEX_ROW = () -> null;
```
[the remark this answers is not in the document]
Comment (ibessonov) [reply modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/AbstractPageMemoryIndexStorage.java:200]: I use "null" for "null", so "NO" seems more obvious to me. But I'll take a look at this code and think about renaming, thank you!

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/AbstractPageMemoryIndexStorage.java [hunk 11]
```diff
@@ -192,6 +196,123 @@ public void finishCleanup() {
...
+
+        /**
+         * Maps value from the index tree into the required result.
+         */
+        protected abstract R map(V value);
+
+        /**
+         * Check whether the passed value exceeds the upper bound for the scan.
+         */
+        protected abstract boolean halt(V value);
```
[the remark this answers is not in the document]
Comment (ibessonov) [reply modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/AbstractPageMemoryIndexStorage.java:235]: "uppder boun"?
What's wrong with "halt", that's a common name for stopping something

[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/AbstractPageMemoryIndexStorage.java:239]: I tried to find in our project at least one use of such a name or similar to it - there is none.

**Review by tkalkirill — APPROVED** [review]

**Pull request merged**